Add DNS challenge support for wildcard certificates

hydrogen-mvm opened the pull request:
The following PR adds support for DNS-01 challenges and for building on Windows (MSVC 2022 with OpenSSL 3.4.1).

The changes were designed to maintain source-level compatibility with existing codebases that use this library for the HTTP-01 challenge. Only one optional parameter was added to the init() function to specify the desired challenge type. The callback function prototype remains the same.

Let me know if there are any regressions or changes requested. Thanks for making this excellent, extremely useful library!

hydrogen-mvm committed:
Add support for DNS challenges

The logic was already 99% present in the library.

hydrogen-mvm committed:
Windows fixes

These fixes allow this library to be built on Windows.

hydrogen-mvm committed:
Add DNS challenge instructions

hydrogen-mvm committed:
Fix compilation error

hydrogen-mvm committed:
Add error message

This will throw a more-helpful error message if someone tries to request a wildcard certificate with a HTTP challenge.

jmccl commented:
Thanks for the PR. I'm busy until next week. I'll take a look then.

jmccl reviewed (changes requested):
This throws in the case in which an account has already passed a challenge within the past 300 days. In other words, there are normal cases in which you don't have to pass a challenge and that's not an error. (See the comment just above this code.)

I think this check should just be removed.

(Edit: I messed this up somehow; it should be in respect of the "Add error message" commit. It'll take me a while to work through this and test it. I'll put a comment in when I'm done.)

jmccl commented on lib/acme-lw.cpp:
```diff
@@ -644,12 +646,19 @@ struct AcmeClientImpl
         }
 
+        // This will typically trigger if a HTTP challenge was requested for a wildcard certificate (which only offers the DNS challenge)
+        if (!challenge_found)
```
This check should be removed. There are cases in which a challenge doesn't need to be passed (see comment at about line 600) and this will throw even though a certificate will be issued if execution continues.

jmccl replied:
I saw you changed the check.

hydrogen-mvm committed:
Merge branch 'jmccl:master' into dns-01

hydrogen-mvm committed:
Fix regression

Fix regression where this throws if a challenge isn't requested.

hydrogen-mvm commented:
Thanks for catching that! I did read that comment you put in the code but it didn't occur to me at the time when I slapped that check in.

I looked more closely at the JSON returned by LE and there's a special "wildcard" key that appears if a wildcard certificate is requested, so I now check for that key if a HTTP challenge is requested before throwing that error. If this still regresses, then I will remove the check completely.

MSVC was also issuing hints that some of the JSON variable assignments should be turned into references to avoid making copies of the json trees, so I also changed a couple of variable declarations. These shouldn't cause any issues from the testing I've done, but I can revert those too if requested.

hydrogen-mvm committed:
Update README

jmccl commented:
How are you building this on Windows? Are you using cmake?

Edit: I managed to get it to build and run.

jmccl commented on lib/acme-lw.h:
```diff
@@ -61,29 +61,39 @@ class AcmeClient
     typedef void (*Callback) (  const std::string& domainName,
-                                const std::string& url,
-                                const std::string& keyAuthorization);
+                                const std::string& url,                         // [HTTP] URL of the GET request; [DNS] record name of the TXT record
```
Can you change 'url' to 'key'? 'url' doesn't make much sense in the case of DNS.

hydrogen-mvm replied:
Not sure if you noticed but the callback function is the exact same for both HTTP and DNS challenges. It is "url" for HTTP challenges, and even the keys for DNS challenges look like URLs (_acme-challenge.domain.tld). If that parameter is to be renamed, we'd need something else than "url" or "key".

jmccl replied:
Oh, it hadn't occurred to me that the TXT record was actually a valid url. Hmm. OK, I take your point.

jmccl commented on main/main.cpp:
```diff
@@ -23,10 +23,21 @@ void   writeFile(const string& fileName, const string& contents);
-            << "\t" << url << "\n\nrespond with this\n\n"
-            << "\t" << keyAuthorization << "\n\n"
-            << "Hit any key when done";
+    if (url.find("_acme-challenge.") == 0)
```
There's no way to get to this code at the moment. I'm fine if you revert it. I'll (likley add some code to deal with DNS and the staging environment once you're merged.)

(In addition, I'm not keen on the use of "_acme-challenge" implementation leaking out here.)

hydrogen-mvm replied:
I'll revert main.cpp back to the master version and leave it to you to implement the DNS logic.

I agree that leaking the acme-challenge implementation is not ideal, but unless we change the callback function to take a 4th parameter (the challenge enum value), I'm not sure if there's a better way to detect DNS vs HTTP challenge from the callback parameters alone.

jmccl reviewed (commented):
Mostly minor stuff. (Before merge I'll ask you squash commits.)

jmccl commented on lib/acme-lw.cpp:
```diff
@@ -619,10 +620,17 @@ struct AcmeClientImpl
             {
-                auto challenges = authz.at("challenges");
+                // If the user requests a wildcard certificate but wants to complete a HTTP challenge, then throw an error:
+                if (chg == AcmeClient::Challenge::HTTP && authz.contains("wildcard") && authz.at("wildcard") == true)
```
This change has a similar issue to the previous check. If a wildcard cert is issued (via DNS challenge) it's good for 300 days. That means the same account can re-issue without a challenge. This check prevents that if the desired challenge is HTTP. Now, if there were a challenge it would fail. But if there isn't one, it succeeds. I understand what you're doing, but I don't see a need to be more strict than Let's Encrypt.

hydrogen-mvm replied:
The error message thrown by LE if you request a HTTP challenge for a wildcard certificate is archaic and unhelpful. The library simply throws an error saying that the status is not "valid" (or something among those lines, I don't recall the exact error message). If an end user saw that, they would have absolutely no idea what they did wrong.

The reason is that none of those if () statements that invoke the callback function are executed, so the library loops through all the challenge options, does nothing (basically a no-op), and then requests LE for a certificate.

I will completely remove this check and leave it to you to add your own check if you can think of a better fix. You should test issuing a wildcard cert with a HTTP challenge on your end to see what I mean.

hydrogen-mvm committed:
Revert back main.cpp

As requested

hydrogen-mvm committed:
Remove unwanted check

hydrogen-mvm commented:
- Reverted back main.cpp so you can implement a better way to support DNS challenges there.

- Removed the check for requesting a wildcard certificate with a HTTP challenge. Test it yourself and confirm if the exception that is thrown is acceptable enough for you.

The only thing I didn't do is change the url parameter to "key" but if you're absolutely sure that you want to use "key" then I will make that change in the next commit.

jmccl commented on lib/acme-lw.cpp:
```diff
@@ -604,7 +604,8 @@ struct AcmeClientImpl
         auto json = nlohmann::json::parse(response);
-        auto authorizations = json.at("authorizations");
+        auto& authorizations = json.at("authorizations");
+        bool challenge_found = false;
```
Can you delete the 'challenge_found' variable and uses as it's unused now?

hydrogen-mvm replied:
Thanks for catching that! I removed it in the next commit.

hydrogen-mvm committed:
Remove unused boolean variable

Whoops

jmccl commented:
Would you be kind enough to squash the commits with a commit message of "Add DNS challenge and Windows support"?

hydrogen-mvm commented:
Because I merged the staging fix in the middle of everything, the squashing process is a little messy. Looking to see if there's a better way other than creating another branch and opening a new PR.

jmccl commented:
Whatever's easiest.

hydrogen-mvm committed:
Merge branch 'jmccl:master' into dns-01

hydrogen-mvm commented:
In the interest of time I just opened PR #19 with the squash commit and will close this PR.
